File: leakyBucket.ts
import { Redis } from "ioredis";

const redisClient = new Redis();

async function rateLimitUsingLeakyBucket(
  userID: string,
  uniqueRequestID: string,
  intervalInSeconds: number,
  maximumRequests: number
): Promise<boolean> {
  // userID can be apikey, location, ip
  const requestCount = await redisClient.llen(userID);

  if (requestCount >= maximumRequests) {
    // drop request
    return false;
  }

  // add request id to the end of the queue
  await redisClient.rpush(userID, uniqueRequestID);

  // handle request
  return true;
}

File: leakyBucket.go
// RateLimitUsingLeakyBucket .
func RateLimitUsingLeakyBucket(userID string, uniqueRequestID string, intervalInSeconds int64, maximumRequests int64) bool {
	// userID can be apikey, location, ip
	requestCount := redisClient.LLen(ctx, userID).Val()
	if requestCount >= maximumRequests {
		// drop request
		return false
	}

	// add request id to the end of queue
	redisClient.RPush(ctx, userID, uniqueRequestID)

	// handle request
	return true
}